src/main/java/ch/njol/skript/expressions/ExprVehicle.java

```diff
@@ -18,186 +18,149 @@
-		return "vehicle";
-	}
-	
+
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
+	public Class<?>[] acceptChange(ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
+			if (isDefault() && getParser().isCurrentEvent(VehicleExitEvent.class, EntityDismountEvent.class)) {
+				Skript.error("Setting the vehicle during a dismount/exit vehicle event will create an infinite mounting loop.");
+				return null;
+			}
 			return new Class[] {Entity.class, EntityData.class};
 		}
 		return super.acceptChange(mode);
 	}
-	
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
-			assert delta != null;
-			final Entity[] ps = getExpr().getArray(e);
-			if (ps.length == 0)
+			// The player can desync if setting an entity as it's currently mounting it.
+			// Remember that there can be other entity types aside from players, so only cancel this for players.
+			Predicate<Entity> predicate = Player.class::isInstance;
+			if (event instanceof EntityMountEvent && predicate.test(((EntityMountEvent) event).getEntity())) {
+				return;
+			}
+			if (event instanceof VehicleEnterEvent && predicate.test(((VehicleEnterEvent) event).getEntered())) {
+				return;
+			}
+			Entity[] passengers = getExpr().getArray(event);
+			if (passengers.length == 0)
 				return;
-			final Object o = delta[0];
-			if (o instanceof Entity) {
-				((Entity) o).eject();
-				final Entity p = CollectionUtils.getRandom(ps);
-				assert p != null;
-				p.leaveVehicle();
-				((Entity) o).setPassenger(p);
-			} else if (o instanceof EntityData) {
-				for (final Entity p : ps) {
-					final Entity v = ((EntityData<?>) o).spawn(p.getLocation());
-					if (v == null)
+			assert delta != null;
+			Object object = delta[0];
+			if (object instanceof Entity) {
+				Entity entity = (Entity) object;
+				entity.eject();
+				for (Entity passenger : passengers) {
+					// Avoid infinity mounting
+					if (event instanceof VehicleExitEvent && predicate.test(passenger) && passenger.equals(((VehicleExitEvent) event).getExited()))
 						continue;
-					v.setPassenger(p);
+					if (event instanceof EntityDismountEvent && predicate.test(passenger) && passenger.equals(((EntityDismountEvent) event).getEntity()))
+						continue;
+					assert passenger != null;
+					passenger.leaveVehicle();
+					entity.addPassenger(passenger);
+				}
+			} else if (object instanceof EntityData) {
+				EntityData<?> entityData = (EntityData<?>) object;
+				for (Entity passenger : passengers) {
+					// Avoid infinity mounting
+					if (event instanceof VehicleExitEvent && predicate.test(passenger) && passenger.equals(((VehicleExitEvent) event).getExited()))
+						continue;
+					if (event instanceof EntityDismountEvent && predicate.test(passenger) && passenger.equals(((EntityDismountEvent) event).getEntity()))
+						continue;
```

Comment:
Do you think it would be more efficient to save a boolean variable for the `event instanceof X` checks rather than doing them every time in the loop?

Reply:
It's not in a loop. It's in a function. Each event is also a different class.